File: QueueSystem/QueueModule.py
from logging import exception


class Queue:
    def __init__(self):
        self.q = []
    
    def head(self):
        if not self.isEmpty():
            return self.q[-1]
        else:
            print(f"message in QueueModule: head failed!")

    def enqueue(self, obj):
        self.q.insert(0, obj)

    def dequeue(self):
        if not self.isEmpty():
            obj  = self.q.pop(-1)
            return obj
        else:
            print(f"message in QueueModule: Pop failed!")
            raise IndexError("Queue is Empty")       
                 
    def isEmpty(self):
        return len(self.q) == 0

class QueueEx(Queue):
    def __init__(self, maxSize: int):
        super().__init__()

        if maxSize <= 0:
            raise exception("Size insert a Positive value.")

        self.maxSize = maxSize
        self.size = 0

    def enqueue(self, obj):
        if self.size < self.maxSize:
            super().enqueue(obj)
            self.size += 1
        else:
            print(f"Queue is Full, {obj} was not inserted.")
    
    def dequeue(self):
        return super().dequeue()

    def isEmpty(self):
        return super().isEmpty()

File: QueueSystem/queueModule.R
#' Creates a queue of maximum size - @size
create_queue <- function(size) {
  q <- list()
  q <- append(q, size)
  return(q)
}

#' Inserts an Element to the back of the queue.
#' new q is returned.
#' If queue is full, a message is printed (same queue is returned)
enqueue <- function(q, element) {
  q_len <- length(q) - 1
  max_len <- as.numeric(q[1])
  # cat("queue len:", q_len, end="\n")
  # cat("queue max len:", as.numeric(max_len), end="\n")
  if (q_len < max_len)
    return(append(q, element))
  else
    cat("Enqueue Error: Queue is FULL -", q_len, "\\", max_len, "\n")
    return(q)
}

#' Return the Element, at the queues FRONT.
#' If Queue is empty, return -1.
peek <- function(q) {
  q_len <- length(q) - 1
  if (q_len == 0) {
    cat("Error! Queue is Empty.\n")
    return(NULL) # do not return -1 (The Queue is an Integer queue)
  }
  else
    return(as.numeric(q[2]))
}

#' Returns A new queue after REMOVING the front element.
#' If the queue is empty, the same queue object is returned.
pop <- function(q) {
  q_len <- length(q) - 1
  if (q_len == 0) {
    q_new <- q
    cat("Pop Error: Queue is Empty.\n")
  }
  else{
    if (q_len == 1)
      q_new <- q[1]
    else # no. of elements > 1
      q_new <- append(q[1], q[2:q_len + 1])
  }
  return(q_new)
}

#' Prints the elements in queue and its size.
prnt_queue <- function(q) {
  q_len <- length(q) - 1
  if (q_len == 0)
    cat("Queue is empty, Nothing to print.\n")
  else{
    cat("Queue(Size=", length(q) - 1, "): ", sep = "")
    for (i in 2:length(q)) {
      cat(as.numeric(q[[i]]), ", ", sep = "")
    }
    cat("\n")
  }
}

#' Returns True if the queue is empty and False otherwise.
is_empty <- function(q) {
  q_len <- length(q) - 1
  if (q_len == 0)
    return(TRUE)
  else
    return(FALSE)
}

#' Returns True if the queue is full and False otherwise.
is_full <- function(q) {
  q_len <- length(q) - 1
  max_len <- as.numeric(q[1])
  if (q_len == max_len)
    return(TRUE)
  else
    return(FALSE)
}

#' Returns the current size of the queue.
get_queue_size <- function(q) {
  return(length(q) - 1)
}